Index: gnuradio-core/src/lib/general/gr_clock_recovery_mm_ff.cc
===================================================================
--- gnuradio-core/src/lib/general/gr_clock_recovery_mm_ff.cc	(revision 8301)
+++ gnuradio-core/src/lib/general/gr_clock_recovery_mm_ff.cc	(working copy)
@@ -130,7 +130,7 @@
     }
   }
 
-  consume_each (ii);
+//  consume_each (ii);
 
-  return noutput_items;
+  return ii; //noutput_items;
 }
Index: mblock/src/lib/mb_message.cc
===================================================================
--- mblock/src/lib/mb_message.cc	(revision 8301)
+++ mblock/src/lib/mb_message.cc	(working copy)
@@ -27,7 +27,7 @@
 #include <pmt_pool.h>
 
 static const int CACHE_LINE_SIZE = 64;	// good guess
-static const int MAX_MESSAGES =  1024;	// KLUDGE max number of messages in sys
+static const int MAX_MESSAGES =  0;	// KLUDGE max number of messages in sys
 					//   0 -> no limit
 #if MB_MESSAGE_LOCAL_ALLOCATOR
 
